### PR TITLE
8313800: AArch64: SA stack walking code having trouble finding sender frame when invoking LambdaForms is involved

### PR DESCRIPTION
**Pull request by theRealAph**
This PR fixes a specific problem caused by using r29/rfp to unwind Java code. For some time we have treated r29 as a callee-saved scratch register, and it is freely used by C2-generated code. Therefore, any code in SA that uses getFP() in a compiled frame is very likely to come to grief.

I believe this is the root cause of 8313800, but it's very hard to prove that because because it's something of an intermittent fault.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313800](https://bugs.openjdk.org/browse/JDK-8313800): AArch64: SA stack walking code having trouble finding sender frame when invoking LambdaForms is involved (**Bug** - P4)

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15624/head:pull/15624` \
`$ git checkout pull/15624`

Update a local copy of the PR: \
`$ git checkout pull/15624` \
`$ git pull https://git.openjdk.org/jdk.git pull/15624/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15624`

View PR using the GUI difftool: \
`$ git pr show -t 15624`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15624.diff">https://git.openjdk.org/jdk/pull/15624.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15624#issuecomment-1710529160)

### CODE REVIEW

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/aarch64/AARCH64Frame.java [hunk 1]
```diff
@@ -327,29 +327,22 @@ private Frame senderForEntryFrame(AARCH64RegisterMap map) {
     return fr;
   }
 
+  //------------------------------------------------------------------------------
+  // frame::adjust_unextended_sp
```
Comment (plummercj) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/aarch64/AARCH64Frame.java:331]: This is just repeating the comment below.

**Review by plummercj — COMMENTED** [review]
Overall the changes look good. Have you tested to see if this fixes [JDK-8276210](https://bugs.openjdk.org/browse/JDK-8276210) and [JDK-8248675](https://bugs.openjdk.org/browse/JDK-8248675)? You'll need to remove them from the problem list to test.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/aarch64/AARCH64Frame.java (hunk 1 above)
Comment (theRealAph) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/aarch64/AARCH64Frame.java:331]: Thanks.

**Comment by plummercj** [comment]
> Overall the changes look good. Have you tested to see if this fixes [JDK-8276210](https://bugs.openjdk.org/browse/JDK-8276210) and [JDK-8248675](https://bugs.openjdk.org/browse/JDK-8248675)? You'll need to remove them from the problem list to test.

Actually I'm now unable to reproduce these two issues even without your fix. So I think you can just push your fix and I'll file a new CR to remove the tests from the problem lists (and also close those two CRs as CNR).

**Review by plummercj — APPROVED** [review]

**Comment by RealFYang** [comment]
Neither can I reproduce this issue on linux-riscv64 platform with jdk/jdk tip. But it has the same settings for the x8/fp register and context, I guess it also bears the same problem. 
```
diff --git a/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/riscv64/RISCV64Frame.java b/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/riscv64/RISCV64Frame.java
index 90d070112db..b21a163e970 100644
--- a/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/riscv64/RISCV64Frame.java
+++ b/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/riscv64/RISCV64Frame.java
@@ -322,24 +322,17 @@ private Frame senderForEntryFrame(RISCV64RegisterMap map) {
   //------------------------------------------------------------------------------
   // frame::adjust_unextended_sp
   private void adjustUnextendedSP() {
-    // If we are returning to a compiled MethodHandle call site, the
-    // saved_fp will in fact be a saved value of the unextended SP.  The
-    // simplest way to tell whether we are returning to such a call site
-    // is as follows:
+    // Sites calling method handle intrinsics and lambda forms are
+    // treated as any other call site. Therefore, no special action is
+    // needed when we are returning to any of these call sites.

     CodeBlob cb = cb();
     NMethod senderNm = (cb == null) ? null : cb.asNMethodOrNull();
     if (senderNm != null) {
-      // If the sender PC is a deoptimization point, get the original
-      // PC.  For MethodHandle call site the unextended_sp is stored in
-      // saved_fp.
-      if (senderNm.isDeoptMhEntry(getPC())) {
-        raw_unextendedSP = getFP();
-      }
-      else if (senderNm.isDeoptEntry(getPC())) {
-      }
-      else if (senderNm.isMethodHandleReturn(getPC())) {
-        raw_unextendedSP = getFP();
+      // If the sender PC is a deoptimization point, get the original PC.
+      if (senderNm.isDeoptEntry(getPC()) ||
+          senderNm.isDeoptMhEntry(getPC())) {
+        // DEBUG_ONLY(verifyDeoptriginalPc(senderNm, raw_unextendedSp));
       }
     }
   }
```

**Comment by theRealAph** [comment]
> Neither can I reproduce this issue on linux-riscv64 platform with jdk/jdk tip. But it has the same settings for the x8/fp register and context, I guess it also bears the same problem.

Yes, I'm sure it does.

**Pull request closed without merging**